### ISSUE TITLE
Allow SwitchProducer case-EDProducers to produce different transient products

### ISSUE THREAD

**Issue by makortel** [makortel/framework#451]
Currently SwitchProducer requires all case-EDProducers to produce the same output products. A case where some producer would produce different set of products can be worked around by using EDAliases (as is done in HLT at the moment).

With the implicit/automated device-to-host transfers for Alpaka EDProducers, introduced in https://github.com/cms-sw/cmssw/pull/39428, that rely on the `Transformer` module ability underneath, this constraint becomes quite annoying, because
* Serial backend EDModule registers only the host memory space branches, whereas all device or asynchronous backend EDModules register branches for both host and device memory spaces
* In general the device memory space branches can have different C++ types between the different device backend EDModules

In the "alternative module loading" approach these constraints are not present, although we want to have a policy for the different module variants to produce the same set of persistable (as defined in `classes_def.xml`) data products. In the future it would be nice to be able to enforce that policy, but that goes beyond this issue.

The bottom line is that we should, I think, relax the SwitchProducer constraints such that the case-EDProducers can have differences in their sets of transient data product types.

**Comment by makortel** [comment]
The consistency check for SwitchProducers is done in
https://github.com/cms-sw/cmssw/blob/ed1fb5f8d839181e1d3644e6ad55ef38d61b37b2/FWCore/Framework/src/Schedule.cc#L236-L311

**Comment by wddgit** [comment]
Thanks. I will start looking at this as soon as I've got the JetCorrector work completed.

**Comment by wddgit** [comment]
@makortel Do you think this is simply a matter of modifying the consistency check you referenced above or is there something beyond that you think needs to be modified? Or you don't know and my job is to answer that question?

So far I'm just reading code. Before this project, I have not spent much time looking at SwitchProducer. I am low on the learning curve still, but it is interesting.

**Comment by makortel** [comment]
I think it should be simply a matter of modifying the consistency check, and enhancing the tests to cover the new behavior (i.e. two case-EDProducers can produce different products as long as the differences are transient).

Or, I can't think of there would be anything else in play. But I'd be "happy" to be proven wrong by tests.

**Comment by wddgit** [comment]
Sounds good. Thanks.

**Comment by wddgit** [comment]
Pull Request 40048 resolves this issue

**Comment by makortel** [comment]
Currently SwitchProducer `foo` creates aliases for all products of the chosen case `foo@<case>`. As part of the case product set consistency check relaxation, it could make sense to limit also the aliases to the common set of products.

An example probably makes it more clear. Say we have
```
process.foo = SwitchProducer(
  cpu = EDProducer("ProducerCPU"),
  cuda = EDProduce("ProducerCUDA"),
)
```
where
* `ProducerCPU` produces products (with "friendly class name")
  * `prod`, persistable
  * `prodTr1`, transient
  * `prodTr2`, transient
* `ProducerCUDA` produces products
  * `prod`, persistable
  * `prodTr1`, transient

Earlier this setup would have been an error because the two cases produce different products. Now it would be ok, because the differences are in transient products. In addition, regardless of the chosen case, the following aliases would be set up for the switch
* `prod_foo` -> `prod_foo@<chosen case>`
* `prodTr1_foo` -> `prodTr1_foo@<chosen case>`

i.e. there would never be a(n alias) product `prodctTr2_foo`.

**Comment by wddgit** [comment]
Can one retrieve a product using the label prodTr2_foo@<chosen_case>? I understand it will not be persistent, but is it available in the current process?

**Comment by makortel** [comment]
> Can one retrieve a product using the label prodTr2_foo@<chosen_case>? I understand it will not be persistent, but is it available in the current process?

Currently the `prodTr2_foo@<chosen_case>` is technically available for consumption, and I would keep it that way (or, I don't think it is worth of the effort to try to prevent it).

**Comment by wddgit** [comment]
Not creating aliases for the transient products not produced in all cases is more difficult than the first thing requested in this issue. The obvious solution is not calling "produces" in those cases in the C++ constructor of the SwitchProducer, but that is called in a callback before we know what products will be produced in all cases. The module constructors of other cases might not be called at that point.

I'm sure this is possible, but it does not look easy. I am continuing on this.

**Comment by makortel** [comment]
Good point. To me making the `SwitchProducer` constructor (or really the `callWhenNewProductsRegistered()` callback there) aware of the set of data products common for all cases sounds too complicated (I'd think it would require some mechanism to call the `SwitchProducer` constructors after all other module constructors have been called, or something).

The main (only?) purpose of avoiding the `prodTr2_foo` -> `prodTr2_foo@<chosen case>` alias is to catch (configuration) errors earlier. Cases like a configuration contains a module consuming `prodTr2_foo`, and the test machine leads to `cpu` case being chosen, and the configuration succeeds to run. Then a later test on a machine with `cuda` chosen case would fail.

From this perspective it _might_ be sufficient to let the `prodTr2_foo` branch to exist, but not create the alias to `prodTr2_foo@cpu`. The consumption of the data product should fail even in the `cpu` case, even though the exception message would likely be confusing. Given that we'd like to move on from SwitchProducer (although likely only after 2023 data taking), I would not want to put too much effort here.

**Comment by makortel** [comment]
Done in https://github.com/cms-sw/cmssw/pull/40104
